### ISSUE TITLE
Make it easier to traverse the frame stack for third party tools.

### ISSUE THREAD

**Issue by markshannon** [python/cpython#100987]
Profilers and debuggers need to traverse the frame stack, but the layout of the stack is an internal implementation detail.
However can make some limited promises to make porting tools between Python versions a bit easier.

In order to traverse the stack, the offset of the `previous` pointer needs to be known. To understand the frame, more information is needed.

@pablogsal 
@Yhg1s 
expressed interest in this.

### Linked PRs
* gh-100988
* gh-102640
* gh-103010
* gh-105727

**Comment by markshannon** [comment]
Initially, I propose to refactor the `PyInterpreterFrame` struct such that it starts:

```
typedef struct _PyInterpreterFrame {
    PyCodeObject *f_code;
    struct _PyInterpreterFrame *previous;
    ...
```

Currently `f_code` must be a code object, but we could generalize it to allow other objects.
For example, the shim frame inserted on entry to `_PyEval_EvalFrameDefault` could have that field set to `None` indicating it should be skipped in tracebacks, etc.

 The order of `f_code` and `previous` doesn't really matter, but have `f_code` first makes https://github.com/python/cpython/issues/100719 a bit simpler

**Comment by pablogsal** [comment]
Let me collect some feedback from maintainers of debuggers and profilers and will comment here the requirements so we can think of solutions.

**Comment by markshannon** [comment]
@pablogsal Any feedback?

**Comment by markshannon** [comment]
We can further improve traversal of the `_PyInterpreterFrame` for debugging and introspection by allowing C extensions to create frames without the rigmarole of creating a code object.

We should rename the `f_code` field to `f_executable`, and allow any object.

```C
typedef struct _PyVMFrame {
    PyObject *f_executable;
    struct _PyVMFrame *previous;
} PyVMFrame;
```
Although tools and the VM should tolerate any object, we should in practice only allow a few classes:
* CodeObject: Implies that the `PyVMFrame` is a full `_PyInterpreterFrame`. Only the VM should make this kind of frame
* Builtin function, method descriptor, slot wrapper, etc. The frame represents a call to the given object.
* `None`: Internal shim. Tools should skip this frame.
* Tuple: First three items should be `name`, `filename`, `flags` where `flags` determine the meaning of additional entries.

The tuple form is for tools like Cython, Nanobind, etc. Creating a tuple of `str`s and `int`s is much simpler and faster than creating a fake code object.

C extension can link themselves into the frame stack at the cost of about 4 memory writes, and 3 reads:
```
    PyVMFrame frame;
    frame.previous = tstate->current_frame.frame;
    frame.f_executable = &EXECUTABLE_OBJECT;
    tstate->current_frame.frame = &frame;
    /* body of function goes here */
    tstate->current_frame.frame = frame.previous;
```

We can do this for builtins functions by modifying the vectorcall function assigned to the builtin function/method descriptor.
We would need to benchmark this to see the performance impact, but it will be much cheaper than `sys.activate_stack_trampoline()`

**Comment by pablogsal** [comment]
> @pablogsal Any feedback?

I have reached out again to tool authors, give me a couple of days to gather comments. Apologies for the delay

**Comment by itamarst** [comment]
@benfred -^

**Comment by markshannon** [comment]
I've made a branch that adds "lightweight" frames (just a pointer to a "code" object and a link pointer), and inserts one for each call to a builtin function in the interpreter. The performance impact is negligible and all builtin function and class calls are present in the frame stack.

Branch: https://github.com/python/cpython/compare/main...faster-cpython:cpython:allow-non-python-frames?expand=1

Performance: https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20230315-3.12.0a6%2B-3e2c3ab/bm-20230315-linux-x86_64-faster%252dcpython-allow_non_python_fra-3.12.0a6%2B-3e2c3ab-vs-base.png

**Comment by markshannon** [comment]
It's a proof of concept, it was easier to remove `_PyFrame_IsEntryFrame` than re-implement it.
`_PyFrame_IsEntryFrame` can be added back, if necessary.

**Comment by carljm** [comment]
We are also very interested in this proposal from the Cinder JIT perspective.

One difference I see with our use case compared to what the draft PR so far aims to support is that we would like to be able to link in "minimal frames" that are still considered "complete": they are fetched by `PyEval_GetFrame()` and can be materialized into a full `PyFrameObject`. (In the draft PR here, only `_PyInterpreterFrame` frames are considered "complete".) We don't want to constantly keep a `_PyInterpreterFrame` (`localsplus` contents) up to date while the JIT is running (this is expensive), so we'd need to get some kind of callback to reify our minimal frame into a `PyFrameObject` on-demand (i.e. some hook into `_PyFrame_GetFrameObject`).

In the tuple form of `f_executable`, could there be a well-known bit-flag in the third element that tells the interpreter "this should be considered a 'complete' frame, and the next word in its struct is a function pointer that will take the VMFrame struct and return a PyFrameObject."?

**Comment by pablogsal** [comment]
> We should rename the `f_code` field to `f_executable`, and allow any object.

In general the feedback is that this will make introspection tool much harder to implement. Currently the fact that this can only be a code object makes it very easy to traverse the frame stack. If you allow **any** Python object it makes it harder or in some cases even impossible. 

If we restrict this to a finite set of possibilities, it still makes it much harder but if we add some kind of enumeration to the frame that tells the tool what's going to be there it makes it a bit easier.

In general I don't think that this proposal helps introspection tools, it actually makes the implementation harder and less efficient because more pointers need to be copied and more logic needs to be included.

**Comment by pablogsal** [comment]
Some comments from authors:

>I feel it won't be too easy to decipher the type of the object remotely. This would likely increase the number of private structures that we need to copy over from Python headers to parse this information (e.g. tuples), making things more complex. Of course  one could just try treating the object as a PyCodeObject and check for failures, but this would now imply a potential loss of captured information, unless all the other object types that can appear here are also handled. Perhaps an extra int field that specifies the type of the object being passed with f_executable might help in this direction, to some extent. But perhaps one simplification that depends on a positive answer to the following question could be adopted: is the value f_executable crucial for the actual execution, or is it just added to carry the frame's metadata (e.g. filename, function name, line number, ...)? If that is added just for the metadata, perhaps that could be added directly to the _PyVMFrame structure in the form of extra fields? There could be a core set of fields that are common to all object types (filename, function qualname, location data), plus a generic PyObject reference that can be consumed easily by in-process tools. However, I can see the downside being that the cost would probably end up being slightly more than just 4 W and 3 R operations in general.

**Comment by pablogsal** [comment]
In general the sentiment is that the more regular the structure is, the easier is for profilers and debuggers to traverse the stack. The more variations and python-isms (as in, using `PyObject*` instead of C structs) the harder it makes for these tools to properly traverse the stack, which goes against (partially) what we are trying to do here

**Comment by markshannon** [comment]
Feedback from who? Which operations for which tools become harder?
It is hard to take vague and anonymous feedback seriously.

No one is forcing tools to handle all possible frames. They can skip frames that have "executable" objects other than code objects. The presence of additional information that tools ignore cannot be worse than that information not being present in the first place.
Add not all tools will ignore it; the PR already give better tracebacks in the faulthandler module.

> In general the sentiment is that the more regular the structure is, the easier is for profilers and debuggers to traverse the stack. The more variations and python-isms (as in, using PyObject* instead of C structs) the harder it makes for these tools to properly traverse the stack, which goes against (partially) what we are trying to do here

Two fields, one pointing to the next frame, and one pointing to the executable object, seems quite regular to me.
Traversal of the stack is trivial. Just follow the `previous` pointer.

**Comment by markshannon** [comment]
It might be informative to compare this with adding `perf` support:
* Adding perf frames causes a slowdown of  [8%](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20230323-3.12.0a6%2B-2aab3df). The PR above has negligible performance impact.
* This works on Windows and any machine that does not have `perf` installed.
* It works with PEP 523 or PEP 669.

**Comment by markshannon** [comment]
The question (as far as I am concerned) is this:
What restrictions should be placed on the "executable" object?

Should those be structural, i.e. the start of all "executable" object must have the same layout,
or should they be nominal, only a fixed set of classes would be allowed as the type of the "executable"?

The tighter the restrictions, the easier it is for introspection tools. The looser the restrictions, the easier it is for producers of frame.

**Comment by pablogsal** [comment]
> The purpose of adding perf support is that tools can see the mixed C/Python stack.

No, the purpose is that users can use `perf` with both C and Python frames.

**Comment by markshannon** [comment]
Isn't that the same thing?

**Comment by markshannon** [comment]
So it doesn't mean optional for us and we are stuck with it?
In that case it needs a PEP.

Supporting `perf` in a JIT compiler is going to be *lot* of extra work, with a benefit for a few large corps that have the infrastructure to support perf profiling in production, and a cost (worse performance and/or reliability) for many.

**Comment by carljm** [comment]
We could do some bit-packing in the lower bits to provide an easy-to-read "flag" indicating the frame type without adding any additional bulk to frames, and not much extra cost to frame creation. This is what Cinder shadowframes does today: https://github.com/facebookincubator/cinder/blob/cinder/3.10/Include/internal/pycore_shadow_frame_struct.h#L42-L60

In the minimal form, we can leave the low bit 0 to indicate "normal frame, pointer to code object" and set it to 1 to mean "pointer to something new and different." Then existing tools that want to just handle normal code objects like they already do have only to add a single bit test to discard frames they don't want to deal with.

There are another two bits we could play with if we want to provide streamlined indication of any other common cases (builtin function, maybe?).

I hope we can make life easier for existing inspection tools by making it really easy to detect the common cases they want to care about, but I also hope (from the Cinder JIT perspective) that at least one of the valid options for `f_executable` is "extensible. E.g. if`(name, filename, lineo)` tuple is allowed, that it's also valid to have a longer tuple carrying additional payload, with the first three elements interpreted as `name, filename, lineo`.

**Comment by carljm** [comment]
> Classes

I'm curious, why would we want frames to hold a pointer to a class (I assume while executing the class body?) rather than to the code object of the class body?

**Comment by markshannon** [comment]
> I'm curious, why would we want frames to hold a pointer to a class

```Py
class C: pass

c = C()  # This is a call to a class
```

**Comment by markshannon** [comment]
@pablogsal 
Is checking for five or six distinct values, rather than two or three that big a deal?
Also, why is comparing to an address a problem, whereas comparing to an int is not?
`frame->f_kind == CODE+OBJECT_KIND` is a 32 bit comparison.
`frame->f_executable == &PyCode_Type` is a 64 bit comparison. 
Fetching the address of `PyCode_Type` will need the symbol table, but you'll need that anyway.

**Comment by pablogsal** [comment]
> Fetching the address of `PyCode_Type` will need the symbol table, but you'll need that anyway.

Not necessarily. These tools need to work some times with stripped binaries or core files and requiring the symbol table there can be a huge pain compared with just checking against an integer, as we are vendoring the headers anyway. In particular in core files is a huge pain because the address may not be in the core if is in the `.rodata` segment.

**Comment by markshannon** [comment]
I appreciate that having extra information will make life easier for a few tool authors, but it might make things a little bit slower for very many Python users. 

How do you get the frame without any symbols? 

> If we allow random classes...

Any class outside of the fixed set (whatever that ends up being) should be ignored, so no "random" classes.

**Comment by pablogsal** [comment]
> How do you get the frame without any symbols?
> 
> >

Find the interpreter state and having the headers vendor so we know the offsets. The interpreter state can be found because we (cpython) place the runtime structure in its own section so it can be found without symbols:

https://github.com/python/cpython/blob/7703def37e4fa7d25c3d23756de8f527daa4e165/Python/pylifecycle.c#L100

**Comment by carljm** [comment]
> I don't see the value in tagging bits. The tag you propose holds no additional information, as the same value can be got with the simple comparison `f_executable == &PyCodeObject`

Sure, if you have `&PyCodeObject` available.

Tagging bits could "make life easier for a few tool authors" in the scenarios @pablogsal is mentioning without "making things slower for very many Python users."

**Comment by markshannon** [comment]
Tagging might solve the performance issue. But we need to support 32bit machines, so we only have 2 bits to play with, which is not enough.

**Comment by markshannon** [comment]
If tools can find the runtime, then we can put an array of pointers there. No runtime overhead at the cost of ~40 bytes.
```C
PyObject *callable_types[] = {
   &PyCode_Type,
   ...
};
```

**Comment by markshannon** [comment]
@pablogsal 
Where should the array go, exactly?

**Comment by markshannon** [comment]
@carljm

> I hope we can make life easier for existing inspection tools by making it really easy to detect the common cases they want to care about, but I also hope (from the Cinder JIT perspective) that at least one of the valid options for f_executable is "extensible". E.g. if(name, filename, lineo) tuple is allowed, that it's also valid to have a longer tuple carrying additional payload, with the first three elements interpreted as name, filename, lineo.

I don't see a problem with that.
Tools should check the length of the tuple before extracting the contents, for safety.
We could allow any length array, specifying only that the first three elements, if they exist, should be the name, filename and line number.
`("foo",)` and `("foo", "foo.py", 121, "special-data-34.8")` should both be acceptable.

@pablogsal Would this be OK, or is this too complex for your tastes?

**Comment by P403n1x87** [comment]
> How do you get the frame without any symbols?

Austin uses symbols to locate `_PyRuntime`, but if those are not available, there is a fallback on BSS scan to locate something that looks like _PyRuntime or an interpreter state. So symbols are not _strictly_ required (but good to have of course!).
